### PR TITLE
feat/bug: Diskwriter should notify source node on shutdown

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-764

**Changes:**
- Improved `network/router` shutdown - more robust
- Fixed bug, missing `networkFile.Close()` in encoding pipeline - `disk writer already exists`
- Disk writer node during termination request encoding pipeline close from all related source nodes
- Added `network/router/balancer` and `network/router/pipeline` packages.
image

### CODE REVIEW

**Review by Matovidlo — APPROVED** [review]

File: internal/pkg/service/stream/storage/level/local/diskwriter/network/rpc/pb/networkFile.proto [hunk 1]
```diff
@@ -8,6 +8,7 @@ import "google/protobuf/timestamp.proto";
 
 service NetworkFile {
   rpc Open(OpenRequest) returns (OpenResponse) {}
+  rpc WaitForServerTermination(WaitForServerTerminationRequest) returns (stream ServerIsTerminatingResponse);
   rpc Write(WriteRequest) returns (WriteResponse) {}
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/level/local/diskwriter/network/rpc/pb/networkFile.proto:12]: gRPC has only one direction of calls, `client -> server`, `Source Node -> Disk Writer Node`.

But we need to signal `Disk Writer Node` termination to the `SourceNode` ( there is the pipeline start).

---------

So we use `stream`, the `client - SourceNode` opens an long-lived stream, and `server - Disk Writer Node` can send termination request.

File: internal/pkg/service/stream/storage/level/local/diskwriter/network/router/balancer/balancer.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package router
+package balancer
 
 import (
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/level/local/diskwriter/network/router/balancer/balancer.go:3]: I created new `balancer` and `pipeline` (sub) package, to split responsibilities.

File: internal/pkg/service/stream/dependencies/source.go [hunk 3]
```diff
@@ -91,9 +91,9 @@ func newSourceScope(svcScp ServiceScope, distScp dependencies.DistributionScope,
 		return nil, err
 	}
 
-	d.Plugins().RegisterSinkPipelineOpener(func(ctx context.Context, sinkKey key.SinkKey, sinkType definition.SinkType) (pipeline.Pipeline, error) {
+	d.Plugins().RegisterSinkPipelineOpener(func(ctx context.Context, sinkKey key.SinkKey, sinkType definition.SinkType, onClose func()) (pipeline.Pipeline, error) {
 		if d.Plugins().IsSinkWithLocalStorage(sinkType) {
-			return d.storageRouter.OpenPipeline(ctx, sinkKey)
+			return d.storageRouter.OpenPipeline(ctx, sinkKey, onClose)
 		}
```
Comment (michaljurecko) [inline internal/pkg/service/stream/dependencies/source.go:97]: Close of a pipeline can be now triggered by the disk writer node shutdown, so it is now async,... solved by the `onClose` callbacks.

Comment (michaljurecko) [reply internal/pkg/service/stream/dependencies/source.go:97]: The `onClose` callback notifies the parent layer about the change.

File: internal/pkg/service/stream/storage/level/local/diskwriter/network/router/router_test.go [hunk 4]
```diff
@@ -211,7 +211,6 @@ func TestRouter_ShutdownDiskWriterNodeFirst(t *testing.T) {
 {"level":"info","message":"closing sink router","nodeId":"source","component":"sink.router"}
 {"level":"info","message":"closed sink router","nodeId":"source","component":"sink.router"}
 {"level":"info","message":"stopping storage statistics collector","nodeId":"source","component":"statistics.collector"}
-{"level":"debug","message":"sync done","nodeId":"source","component":"statistics.collector"}                                                                
 {"level":"info","message":"storage statistics stopped","nodeId":"source","component":"statistics.collector"}
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/level/local/diskwriter/network/router/router_test.go:214]: Fixed unstable test, ... statics may be synce by the periodical syncer, so there is no record for sync on shutdown.

**Comment by michaljurecko** [comment]
(rebased above `main`)

**Review by Matovidlo — APPROVED** [review]

File: internal/pkg/service/stream/storage/level/local/diskwriter/network/rpc/client.go [hunk 5]
```diff
@@ -38,18 +46,24 @@ func OpenNetworkFile(ctx context.Context, sourceNodeID string, conn *transport.C
...
-          "timeout": "10s",
-		  "retryPolicy": {
-			  "MaxAttempts": 5,
+		"methodConfig": [
+          {
+		    "name": [{}],
+		    "waitForReady": true,
+            "timeout": "10s",
+		    "retryPolicy": {
+              "MaxAttempts": 5,
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/stream/storage/level/local/diskwriter/network/rpc/client.go:55]: Thanks, I didn't notice that.

[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/stream/storage/level/local/diskwriter/network/rpc/client.go:55]: Fixed in: https://github.com/keboola/keboola-as-code/pull/1986
Commit: https://github.com/keboola/keboola-as-code/pull/1986/commits/67e6a04e98b1b3a93b6e31f3fc595e4bbfe527d0

**Pull request merged**